zebra-consensus/src/error.rs

```diff
@@ -24,8 +28,45 @@ pub enum TransactionError {
     #[error("coinbase input found in non-coinbase transaction")]
     CoinbaseInputFound,
 
+    #[error("coinbase transaction MUST NOT have any JoinSplit descriptions or Spend descriptions")]
+    CoinbaseHasJoinSplitOrSpend,
+
     #[error("coinbase transaction failed subsidy validation")]
     Subsidy(#[from] SubsidyError),
+
+    #[error("transaction version number MUST be >= 4")]
+    WrongVersion,
+
+    #[error("at least one of tx_in_count, nShieldedSpend, and nJoinSplit MUST be nonzero")]
+    NoTransfer,
+
+    #[error("if there are no Spends or Outputs, the value balance MUST be 0.")]
+    BadBalance,
+
+    /// Could not verify a transparent script
+    Script(#[from] zebra_script::Error),
+
+    // XXX change this when we align groth16 verifier errors with bellman
+    // and add a from annotation when the error type is more precise
+    #[error("spend proof MUST be valid given a primary input formed from the other fields except spendAuthSig")]
+    Groth16(BoxError),
+
+    #[error(
+        "joinSplitSig MUST represent a valid signature under joinSplitPubKey of dataToBeSigned"
+    )]
+    Ed25519(#[from] ed25519::Error),
+
+    #[error("bindingSig MUST represent a valid signature under the transaction binding validating key bvk of SigHash")]
+    RedJubjub(redjubjub::Error),
+}
+
+impl From<BoxError> for TransactionError {
+    fn from(err: BoxError) -> Self {
+        match err.downcast::<redjubjub::Error>() {
+            Ok(e) => TransactionError::RedJubjub(*e),
+            Err(e) => TransactionError::Internal(e),
+        }
+    }
 }
```

Comment:
@yaahc this needs some tidying / more impls as you can see.

I'm not if we need/can have a fallback error variant, like an `Internal`, if we fail to downcast a `redjubjub::Error` into a `TransactionError`, or what, since that's not exactly a consensus rule failure in itself, it's our bug?

Reply:
Sounds reasonable, lets go ahead and remove the `Internal` variant completely and just panic if we run into an error that we don't expect. I think the only other source of errors are errors introduced by `tower`, and if we end up running into those we can probably change this to a `TryFrom` impl and handle tower errors separately at the conversion site or figure it out once we encounter it.